tests/test_meta/test_hierarchical_predictor.py

```diff
@@ -0,0 +1,144 @@
+        {"shrinkage": "equal"},
+        {"shrinkage": "relative"},
+        {"shrinkage": "min_n_obs", "min_n_obs": 10},
+        {"shrinkage": "constant", "alpha": 0.5},
+    ],
+)
+def test_fit_predict(meta_cls, base_estimator, task, fallback_method, shrinkage):
+    """Tests that the model can be fit and predict with different configurations of fallback and shrinkage methods if
+    X to predict contains same groups as X used to fit.
+    """
+    X, y, groups = make_hierarchical_dataset(task)
+
+    meta_model = meta_cls(estimator=base_estimator, groups=groups, fallback_method=fallback_method, **shrinkage).fit(
+        X, y
+    )
+
+    assert meta_model.estimators_ is not None
+    assert meta_model.predict(X) is not None
+
+    if task in {"binary-classification", "multiclass-classification"}:
+        assert meta_model.predict_proba(X) is not None
+
+
+@pytest.mark.parametrize(
+    "meta_cls,base_estimator,task",
+    [
+        (HierarchicalRegressor, LinearRegression(), "regression"),
+        (HierarchicalClassifier, LogisticRegression(), "binary-classification"),
+        (HierarchicalClassifier, LogisticRegression(), "multiclass-classification"),
+    ],
+)
+@pytest.mark.parametrize("fallback_method,context", [("raise", pytest.raises(KeyError)), ("parent", does_not_raise())])
+def test_fallback(meta_cls, base_estimator, task, fallback_method, context):
+    """Tests that the model fails or not when predicting with different fallback methods if X to predict contains
+    unseen group values.
+    """
+    X, y, groups = make_hierarchical_dataset(task)
+
+    meta_model = meta_cls(estimator=base_estimator, groups=groups, fallback_method=fallback_method).fit(X, y)
+    X.loc[:, groups] = "unseen_group_value"
+    with context:
+        meta_model.predict(X)
+
+
+@pytest.mark.parametrize(
+    "meta_cls,base_estimator,task,metric",
+    [
+        (HierarchicalRegressor, LinearRegression(), "regression", r2_score),
+        (HierarchicalClassifier, LogisticRegression(), "binary-classification", accuracy_score),
+        (HierarchicalClassifier, LogisticRegression(), "multiclass-classification", accuracy_score),
+    ],
+)
+@pytest.mark.parametrize(
+    "shrinkage",
+    [
+        {"shrinkage": None},
+        {"shrinkage": "equal"},
+        {"shrinkage": "relative"},
+        {"shrinkage": "min_n_obs", "min_n_obs": 10},
+        {"shrinkage": "constant", "alpha": 0.5},
+    ],
+)
+def test_shrinkage(meta_cls, base_estimator, task, metric, shrinkage):
+    """Tests that the model performance is better than the base estimator when predicting with different shrinkage
+    methods.
+    """
+    X, y, groups = make_hierarchical_dataset(task)
+
+    meta_model = meta_cls(estimator=clone(base_estimator), groups=groups, **shrinkage).fit(X, y)
+    base_model = clone(base_estimator).fit(X.drop(columns=groups), y)
+
+    assert metric(y, base_model.predict(X.drop(columns=groups))) <= metric(y, meta_model.predict(X))
```

Comment:
This is some sort of behavioural test

Reply:
I guess there's an opportunity for one more test that leverages the [DummyRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyRegressor.html). 

```python
df = pd.DataFrame([
  {"grp": "a", "x": 1, "y": 0},
  {"grp": "a", "x": 1, "y": 0},
  {"grp": "a", "x": 1, "y": 0},
  {"grp": "b", "x": 1, "y": 2},
  {"grp": "b", "x": 1, "y": 2},
  {"grp": "b", "x": 1, "y": 2},
])
```

We should be able to confirm that the parent model would have a model that always predicts 1. Then depending on the shrinkage factor we might be able to have some expectation on predictions for group a and b, no?

Reply:
For some reason I didn't see this comment. I ended up with a somewhat similar approach, namely: a dataset that could be used for both regression and classification with dummy estimators (see `make_hierarchical_dummy`) used in `test_expected_output` function.

There are comments on why the expected outputs are the way they are. I hope it is good enough. Let me know what you think